tests/test_chains.py

```diff
@@ -0,0 +1,436 @@
+            prompt = prompt + [client.image_to_prompt(init_image)]
+        if negative_prompt is not None:
+            prompt = prompt + [generation.Prompt(text="fish",parameters=generation.PromptParameters(weight=-1.3))]
+
+        step_parameters = dict(
+            scaled_step=0,
+        )
+        if init_noise_scale is not None:
+            sampler=generation.SamplerParameters(init_noise_scale=init_noise_scale)
+            step_parameters['sampler'] = sampler
+        if schedule is not None:
+            # [1.0, 0.01] is the default
+            start_schedule = schedule[0]
+            end_schedule = schedule[1]
+            step_parameters['schedule'] = generation.ScheduleParameters(
+                    start=start_schedule,
+                    end=end_schedule,
+                )
+        parameters = [generation.StepParameter(**step_parameters)]
+
+        if seeds is not None:
+            image_parameters = generation.ImageParameters(
+                seed=seeds,
+                parameters=parameters)
+        else:
+            image_parameters = generation.ImageParameters(
+                parameters=parameters
+            )
+
+        request_id = str(uuid.uuid4())
+        print('generate request_id',request_id)
+
+
+        rq = generation.Request(
+            engine_id=generate_engine,
+            request_id=request_id,
+            prompt=prompt,
+            image=image_parameters,
+            **kwargs
+        )
+        
+        return rq
+
+
+def enhance_request(
+        enhance_engine:str,
+        weight: float = 1.0,
+        init_image: str = None,
+        **kwargs
+    ) -> generation.Request:
+        
+        extras = Struct()
+        extras.update({
+            'weight': weight
+        })
+
+        prompt = []
+        if init_image is not None:
+            prompt = prompt + [client.image_to_prompt(init_image)]
+
+        image_parameters = generation.ImageParameters()
+
+        request_id = str(uuid.uuid4())
+
+        rq = generation.Request(
+            engine_id=enhance_engine,
+            request_id=request_id,
+            image=image_parameters,
+            extras=extras,
+            prompt=prompt,
+            **kwargs
+        )
+        
+        return rq
+
+def FakeChainGenerate(chain: generation.ChainRequest, **kwargs) -> Generator[generation.Answer, None, None]:
+    # Not a full implementation of chaining, but enough to test current api.Context layer
+    artifacts = []
+    for stage in chain.stage:
+        stage.request.MergeFrom(generation.Request(prompt=[generation.Prompt(artifact=a) for a in artifacts]))
+        artifacts = []
+        for answer in stability_api.stub.Generate(stage.request):
+            artifacts.extend(answer.artifacts)
+    for artifact in artifacts:
+        yield generation.Answer(artifacts=[artifact])   
+
+def ChainGenerate(chain_rq: generation.ChainRequest, **kwargs) -> Generator[generation.Answer, None, None]:
```

Comment:
There is also test code to mock GRPC stub and chains at https://github.com/Stability-AI/stability-sdk/blob/main/tests/test_api.py#L26 you could use same MockStub and issue requests to it

Reply:
Thank you for that! Made changes to use the mock stub.